src/docfx/restore/CredentialProvider.cs

```diff
@@ -4,36 +4,92 @@
+        private static readonly ConcurrentDictionary<string, Dictionary<string, string>> s_credentialCache = new();
+
+        private readonly IReadOnlyDictionary<string, HttpConfig> _configCredentials;
+        private readonly Func<string, Task<Dictionary<string, HttpConfig>>>? _getCredential;
+
+        public bool SupportRefreshToken => _getCredential != null;
+
+        public CredentialProvider(
+            IReadOnlyDictionary<string, HttpConfig> configCredentials, Func<string, Task<Dictionary<string, HttpConfig>>>? getCredential = null)
         {
-            _credentials = credentials;
+            _configCredentials = configCredentials;
+            _getCredential = getCredential;
         }
 
-        public Dictionary<string, string> GetCredentials(HttpRequestMessage request)
+        public bool FillInCredentials(HttpRequestMessage request)
         {
+            var credentialUpdated = false;
             if (request.RequestUri?.ToString() is string url)
             {
-                return _credentialCache.GetOrAdd(url, _ =>
+                var credentials = GetCredentials(url);
+
+                foreach (var (key, value) in credentials)
                 {
-                    var credentials = new Dictionary<string, string>();
-                    GetCredentialsCore(credentials, url, _credentials);
-                    GetCredentialsCore(credentials, url, OpsAccessor.FallBackCredentials);
-                    return credentials;
-                });
+                    if (request.Headers.TryGetValues(key, out var originalCredentials))
+                    {
+                        request.Headers.Remove(key);
+                        if (originalCredentials.All(t => t != value))
+                        {
+                            credentialUpdated = true;
+                        }
+                    }
+                    else
+                    {
+                        credentialUpdated = true;
+                    }
+                    request.Headers.Add(key, value);
+                }
             }
-            return new Dictionary<string, string>();
+            return credentialUpdated;
         }
 
-        private static void GetCredentialsCore(Dictionary<string, string> credentials, string url, IReadOnlyDictionary<string, HttpConfig> providedCredentials)
+        public async Task RefreshCredential(HttpRequestMessage request)
+        {
+            if (request.RequestUri?.ToString() is string url && _getCredential != null)
+            {
+                await s_credentialRefreshSemaphore.WaitAsync(CancellationToken.None);
+                if (!FillInCredentials(request))
+                {
+                    var newCredentials = await _getCredential.Invoke(url);
+                    foreach (var (key, value) in newCredentials)
+                    {
+                        s_refreshedCredentials[key] = value;
```

Comment:
>s_refreshedCredentials [](start = 24, length = 22)

seems that we don't need `s_refreshedCredentials` and the semaphore. can we build out a local variable `refreshedCredentials` and update `s_credentialCache` directly?

Reply:
semaphore is used to avoid multiple thread refresh the credential in the same time

---
In reply to: [551896148](https://github.com/dotnet/docfx/pull/6962#discussion_r551896148) [](ancestors = 551896148)